distributed/scheduler.py

```diff
@@ -4377,16 +4392,23 @@ async def remove_worker(self, address, stimulus_id, safe=False, close=True):
             self.bandwidth_workers.pop((address, w), None)
             self.bandwidth_workers.pop((w, address), None)
 
-        def remove_worker_from_events():
-            # If the worker isn't registered anymore after the delay, remove from events
-            if address not in self.workers and address in self.events:
-                del self.events[address]
-
-        cleanup_delay = parse_timedelta(
-            dask.config.get("distributed.scheduler.events-cleanup-delay")
-        )
-        self.loop.call_later(cleanup_delay, remove_worker_from_events)
-        logger.debug("Removed worker %s", ws)
+        # This way, the tornado callback does not hold a reference to the
+        # Scheduler instance
+        # However, the WorkerState objects themselves hold refs to TaskState
+        # We need to ensure this callback is cancelled before closing to ensure
+        # proper GC
+        # _workers = self.workers
+        # _events = self.events
+        # def remove_worker_from_events():
+        #     # If the worker isn't registered anymore after the delay, remove from events
+        #     if address not in _workers and address in _events:
+        #         del _events[address]
+
+        # cleanup_delay = parse_timedelta(
+        #     dask.config.get("distributed.scheduler.events-cleanup-delay")
+        # )
+        # self.loop.call_later(cleanup_delay, remove_worker_from_events)
+        # logger.debug("Removed worker %s", ws)
```

Comment:
tornado `call_later`, `add_callback`, etc. things keep a reference around. With the above construction, we can break references to the scheduler but the sets/dicts still include references to `WorkerState` instances which themselves hold references to `TaskState` objects.